server/bleep/src/agent/symbol.rs

```diff
@@ -0,0 +1,370 @@
+use futures::TryStreamExt;
+
+use crate::agent::{exchange::CodeChunk, Agent};
+use crate::intelligence::{code_navigation::FileSymbols, Language, TSLanguage};
+use crate::llm_gateway;
+use crate::webserver::intelligence::{get_token_info, TokenInfoRequest};
+use anyhow::{Context, Result};
+use tracing::log::{debug, info, warn};
+
+pub struct ChunkWithSymbols {
+    pub chunk: CodeChunk,
+    pub symbols: Vec<Symbol>,
+}
+
+impl Agent {
+    pub async fn extract_symbols(&self, chunk: CodeChunk) -> Result<ChunkWithSymbols> {
+        const MAX_REF_DEFS: usize = 5; // Ignore symbols with more than this many cross-file refs/defs
+
+        // get hoverable elements
+        let document = self
+            .app
+            .indexes
+            .file
+            .by_path(&self.repo_ref, &chunk.path, None)
+            .await?
+            .with_context(|| format!("failed to read path: {}", &chunk.path))?;
+
+        let hoverable_ranges = document
+            .hoverable_ranges()
+            .ok_or_else(|| anyhow::anyhow!("no hoverable ranges"))?;
+
+        let all_docs = {
+            let associated_langs = match document.lang.as_deref().map(TSLanguage::from_id) {
+                Some(Language::Supported(config)) => config.language_ids,
+                _ => &[],
+            };
+            self.app
+                .indexes
+                .file
+                .by_repo(&self.repo_ref, associated_langs.iter(), None)
+                .await
+        };
+
+        // get references and definitions for each symbol
+        let related_symbols = futures::future::join_all(
+            hoverable_ranges
+                .iter()
+                .filter(|range| {
+                    (range.start.byte >= chunk.start_byte) && (range.start.byte < chunk.end_byte)
+                })
+                .map(|range| {
+                    get_token_info(
+                        TokenInfoRequest {
+                            relative_path: chunk.path.clone(),
+                            repo_ref: String::new(), // FIXME
+                            branch: None,
+                            start: range.start.byte,
+                            end: range.end.byte,
+                        },
+                        &self.repo_ref,
+                        self.app.indexes.clone(),
+                        &document,
+                        &all_docs,
+                    )
+                }),
+        )
+        .await;
+
+        // filter references and definitions
+        let mut symbols = related_symbols
+            .into_iter()
+            .filter_map(Result::ok)
+            .zip(hoverable_ranges.into_iter().filter(|range| {
```

Comment:
Can you comment explaining these filters?

Reply:
I added

```
        // filter references and definitions
        // 1: symbol shouldn't be in the same file
        // 2: number of refs/defs should be less than 5 to avoid very common symbols (iter, unwrap...)
        // 3: also filter out symbols without refs/defs
```